File: COTE/Hackerrank/InterviewPreparationKit/Arrays-ArrayManipulation.py
# https://www.hackerrank.com/challenges/crush/problem?isFullScreen=true&h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=arrays
# 2022.03.23

#Failed due to timeout 🧐🧐🧐
def arrayManipulation(n, queries):
    # Write your code here

    graph = [[0]*n for i in range(m)]
    
    for i in range(m):
        start = queries[i][0]
        end = queries[i][1]
        k = queries[i][2]
        # print("---i>",i)
        for x in range(m): #💜
            for j in range(start-1, end):
                graph[x][j]+=k
        # print("graph>>>:",graph)
    
    return findMaxVal(graph)
    
def findMaxVal(arr):
    length=len(arr)
    maxVal=0
    for i in range(length):
        maxVal = max(maxVal, max(arr[i]))
    return maxVal


#Improved version... https://youtu.be/6dbfeaqRBFU
#[ ]
def arrayManipulation(n, queries):
    # Write your code here
    graph = [0]*(n+1)
    
    for i in range(m):
        # print("i:",i)
        start = queries[i][0]
        end = queries[i][1]
        k = queries[i][2]

        graph[start-1]+=k
        graph[end]-=k
    # print("graph>>>:",graph)
    
    ans = 0
    cur = 0
    for i in graph:
        cur += i
        if cur > ans:
            ans=cur
    return ans

File: COTE/Hackerrank/InterviewPreparationKit/Arrays-ArrayManipulation.js
function arrayManipulation(n, queries) {
   // Write your code here
   // 💜💜💜
   let zeroArr = '0'.repeat(n+1).split('').map((e) => parseInt(e));
   for(let i=0; i<queries.length; i++){
       // console.log("q[",i,"]:",queries[i])
       let startIdx = queries[i][0]
       let endIdx = queries[i][1]
       let addAmount = queries[i][2]
       
       zeroArr[startIdx-1]+=addAmount
       zeroArr[endIdx]-=addAmount
   }
   // console.log(zeroArr)
   let ans = 0
   let cur=0
   for(let i=0; i<n; i++){
      // 💜
       cur += zeroArr[i]
       if (cur>ans){
           ans = cur
       }
   }
   return ans
}


//interesting but more complicated
//https://medium.com/@Kliative/array-manipulation-hacker-rank-javascript-5a702c1ff044